### PR TITLE
Add js messaging framework

### PR DESCRIPTION
**Pull request by marcosholgado**
Task/Issue URL: https://app.asana.com/0/0/1205384572577324/f

### Description
This PR adds new modules to support the new JS messaging

### Steps to test this PR

- [x] Go to `https://frolicking-hamster-6ded3e.netlify.app`
- [x] Filter the logcat by `message`
- [x] You should see `Message processed JsMessage(context=ContentScopeScripts, featureName=helloWorld, method=foo, params={}, id=null)`
- [x] Filter by `chromium` in the logcat
- [x] Click on the request 1 button
- [x] You should see `"response from request1 {}"`
- [x] Click on the request 2 button
- [x] You should see `"response from request2 {"result":"1"}"`
- [x] Click on the request 3 button
- [x] You should see `"response from request3 {"user":"shakyshane"}"`
- [x] In the menu click on the share button
- [x] You should see `"data from onUpdate subscription {}"`

### CODE REVIEW

**Comment by marcosholgado** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3545** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3545" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3545?utm_source=stack-comment).

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt [hunk 1]
```diff
@@ -196,6 +196,9 @@ import com.duckduckgo.downloads.api.DOWNLOAD_SNACKBAR_LENGTH
 import com.duckduckgo.downloads.api.DownloadCommand
 import com.duckduckgo.downloads.api.FileDownloader
 import com.duckduckgo.downloads.api.FileDownloader.PendingFileDownload
+import com.duckduckgo.js.messaging.api.JsMessageHandler
```
Comment (marcosholgado) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:199]: All the changes in this file as for testing purposes, they will all be reverted before merging.

File: package.json [hunk 2]
```diff
@@ -16,7 +16,7 @@
   "dependencies": {
     "@duckduckgo/autoconsent": "^5.1.0",
     "@duckduckgo/autofill": "github:duckduckgo/duckduckgo-autofill#8.2.0",
-    "@duckduckgo/content-scope-scripts": "github:duckduckgo/content-scope-scripts#4.36.0",
+    "@duckduckgo/content-scope-scripts": "github:duckduckgo/content-scope-scripts#09-14-WIP_adding_android_messaging_support",
```
Comment (marcosholgado) [inline package.json:19]: This will point to the right version before merging

File: package-lock.json [hunk 3]
```diff
@@ -10,7 +10,7 @@
       "dependencies": {
         "@duckduckgo/autoconsent": "^5.1.0",
         "@duckduckgo/autofill": "github:duckduckgo/duckduckgo-autofill#8.2.0",
-        "@duckduckgo/content-scope-scripts": "github:duckduckgo/content-scope-scripts#4.36.0",
+        "@duckduckgo/content-scope-scripts": "github:duckduckgo/content-scope-scripts#09-14-WIP_adding_android_messaging_support",
```
Comment (karlenDimla) [inline package-lock.json:13]: Is this expected to be committed?

Comment (karlenDimla) [reply package-lock.json:13]: Ah saw the comment just below.

File: package.json (hunk 2 above)
Comment (karlenDimla) [reply package.json:19]: What is the right version?

File: js-messaging/js-messaging-impl/src/main/java/com/duckduckgo/js/messaging/impl/RealJsMessageHandler.kt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+) : JsMessageHandler {
+
+    private val moshi = Moshi.Builder().add(JSONObjectAdapter()).build()
+    override fun process(message: String, context: String, secret: String, callback: String, webView: WebView) {
+        try {
+            val adapter = moshi.adapter(JsMessage::class.java)
+            val jsMessage = adapter.fromJson(message)
+
+            jsMessage?.let {
+                if (context == jsMessage.context) {
```
Comment (karlenDimla) [inline js-messaging/js-messaging-impl/src/main/java/com/duckduckgo/js/messaging/impl/RealJsMessageHandler.kt:56]: can we lift the checf for` jsMessage.id != null` here?

**Review by karlenDimla — APPROVED** [review]
Works as expected in the steps to test. LGTM, with just one nit.

**Review by joshliebe — APPROVED** [review]
lgtm 👍

File: js-messaging/js-messaging-impl/src/main/java/com/duckduckgo/js/messaging/impl/RealJsMessageHandler.kt (hunk 4 above)
Comment (marcosholgado) [reply js-messaging/js-messaging-impl/src/main/java/com/duckduckgo/js/messaging/impl/RealJsMessageHandler.kt:56]: The check should be done in each plugin. We have two types of messages that we can received, notifications (without id) and requests (with id). Only the latter require a response from us so is up to the plugin to decide if it should return a response or not and you can't build a JsResponse object without id.

**Pull request merged**